### PR TITLE
Provide a better error message on resolve-by-IP failure

### LINKED ISSUE #2655: malformed IP address: when using hostname
*Description*:
>When trying to use a hostname as the address for a dns_resolver envoy fails to start with the error: [2018-02-21 15:55:25.801][1][critical][main] source/server/server.cc:72] error initializing configuration '/config/envoy.yaml': malformed IP address: kube-dns.kube-system.svc.cluster.local

Config line looks like:
...
static_resources:
...
  clusters:
...
     - name: xds_cluster
        connect_timeout: 0.25s
        lb_policy: ROUND_ROBIN
        http2_protocol_options: {}
        hosts: [{ socket_address: { address: "discovery-service", port_value: 8454, resolver_name: "kube-dns.kube-system.svc.cluster.local" }}]
        dns_resolvers: [{ socket_address: { address: "kube-dns.kube-system.svc.cluster.local", port_value: 53 }}]
...

### PR DESCRIPTION
**Pull request by dio**
Provide a better error message on resolve-by-IP failure

This PR provides a better error message if the resolver fails with malformed IP address.

*Risk Level*: Low

*Testing*:
- Unit test

Fixes #2655 

Signed-off-by: Dhi Aurrahman <dio@rockybars.com>

### CODE REVIEW

File: source/common/network/resolver_impl.cc [hunk 1]
```diff
@@ -67,7 +67,17 @@ resolveProtoSocketAddress(const envoy::api::v2::core::SocketAddress& socket_addr
...
   }
-  return resolver->resolve(socket_address);
+
+  try {
+    return resolver->resolve(socket_address);
+  } catch (EnvoyException& e) {
+    // resolver_name.empty() means resolver tries to resolve using IP address.
+    if (resolver_name.empty()) {
+      throw EnvoyException(fmt::format(
+          "{}. Please try to use 'STRICT_DNS' or 'LOGICAL_DNS' cluster type instead", e.what()));
```
Comment (dio) [inline source/common/network/resolver_impl.cc]: Not really sure if this is correct, should we just tell the user to set the `resolver_name` and define another cluster entry with ip address? (seems confusing while I'm typing this).

Comment (zuercher) [reply source/common/network/resolver_impl.cc]: I think the tricky part about an error here is that, in principal, there could be a named resolver that did DNS (or other hostname) lookups. No such resolver ships with Envoy, though. Setting the cluster type bypasses this resolver altogether, which is the currently the only stock way to use DNS names, AFAIK.

Perhaps the addition to the error should be "Consider setting resolver_name or setting cluster type to 'STRICT_DNS' or 'LOGICAL_DNS'."

Comment (mattklein123) [reply source/common/network/resolver_impl.cc]: I wonder whether this type of error checking would be better placed more at the source? Like in the static cluster and the EDS cluster directly? Thoughts?

Comment (dio) [reply source/common/network/resolver_impl.cc]: @zuercher @mattklein123 sorry missed your comments, I tried to pass the `cluster_type` down, hence we will be sure about the current cluster type. What do you think?

File: source/common/network/resolver_impl.cc [hunk 2]
```diff
@@ -54,6 +54,27 @@ InstanceConstSharedPtr resolveProtoAddress(const envoy::api::v2::core::Address&
...
 
+InstanceConstSharedPtr
+resolveProtoAddress(const envoy::api::v2::core::Address& address,
+                    const envoy::api::v2::Cluster::DiscoveryType& cluster_type) {
+  switch (address.address_case()) {
+  case envoy::api::v2::core::Address::kSocketAddress:
+    try {
+      return resolveProtoSocketAddress(address.socket_address());
+    } catch (EnvoyException& e) {
+      if (cluster_type == envoy::api::v2::Cluster::STATIC) {
```
Comment (mattklein123) [inline source/common/network/resolver_impl.cc]: Doesn't this issue also apply to EDS clusters which I think is also where some of the confusion is coming from?

Comment (dio) [reply source/common/network/resolver_impl.cc]: Woops, yes, we need to check for `envoy::api::v2::Cluster::EDS`. Will fix, either here or in a helper function in `Upstream` namespace.

Comment (mattklein123) [reply source/common/network/resolver_impl.cc]: Thanks. I guess I would vote for helper wrapper in `Upstream` namespace.

**Comment by zuercher** [comment]
What if we make the helper a member function of Upstream::ClusterImplBase? Then it would have access to ClusterImplBase::info() and the cluster type. I think that keeps it contained to the context where it makes sense to report this error.

**Comment by mattklein123** [comment]
@zuercher yeah +1

**Comment by dio** [comment]
@zuercher @mattklein123 done as requested.

File: source/common/upstream/upstream_impl.h [hunk 3]
```diff
@@ -412,6 +412,15 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
    */
   void setOutlierDetector(const Outlier::DetectorSharedPtr& outlier_detector);
 
+  /**
+   * Wrapper around Network::Address::resolveProtoAddress() to catch the specific error in relation
+   * with envoy::api::v2::Cluster::DiscoveryType hence we can throw more meaningful message.
+   * @param address supplies the address proto to resolve.
+   * @return pointer to the Instance from a envoy::api::v2::core::Address.
```
Comment (zuercher) [inline source/common/upstream/upstream_impl.h]: Our convention is to list the return type first. So something like:

`@return Network::Address::InstanceConstSharedPtr the resolved address.`

File: source/common/upstream/upstream_impl.h [hunk 4]
```diff
@@ -412,6 +412,15 @@ class ClusterImplBase : public Cluster, protected Logger::Loggable<Logger::Id::u
    */
   void setOutlierDetector(const Outlier::DetectorSharedPtr& outlier_detector);
 
+  /**
+   * Wrapper around Network::Address::resolveProtoAddress() to catch the specific error in relation
+   * with envoy::api::v2::Cluster::DiscoveryType hence we can throw more meaningful message.
```
Comment (zuercher) [inline source/common/upstream/upstream_impl.h]: Perhaps just `Wrapper around Network::Address::resolveProtoAddress() that provides improved error messages based on the cluster's type.`

File: source/common/upstream/upstream_impl.h (hunk 3 above)
Comment (dio) [reply source/common/upstream/upstream_impl.h]: Noted. Thanks! Changed here cf5a97b

File: source/common/upstream/upstream_impl.h (hunk 4 above)
Comment (dio) [reply source/common/upstream/upstream_impl.h]: OK. Thank you for the suggestion. cf5a97b

**Pull request merged**